### PR TITLE
feat: basic object merging

### PR DESCRIPTION
**Pull request by flub**
This has a basic recursive merging of sentry_value_t objects.
Generally values from the source object have preference, however when
the values of a key are objects themselves in both the target and the
source they are recursively merged.

### CODE REVIEW

File: src/sentry_value.c [hunk 1]
```diff
@@ -806,6 +807,38 @@ sentry_value_is_null(sentry_value_t value)
...
+    for (size_t i = 0; i < obj->len; i++) {
+        char *key = obj->pairs[i].k;
+        sentry_value_t src_val = obj->pairs[i].v;
+        sentry_value_t dst_val = sentry_value_get_by_key(dst, key);
+        if (sentry_value_get_type(dst_val) == SENTRY_VALUE_TYPE_OBJECT
+            && sentry_value_get_type(src_val) == SENTRY_VALUE_TYPE_OBJECT) {
+            if (sentry__value_merge_objects(dst_val, src_val)) {
+                return 1;
+            }
+        } else {
```
Comment (Swatinem) [inline src/sentry_value.c:832]: I think you want to check `dst_val` for null here. You are replacing all existing values, whereas in the description you say that existing `dst` props take precedence.

**Review by Swatinem — APPROVED** [review]
implementation/tests do not match the PR description when it comes to which props take precedence. Otherwise this looks good, if overwriting is actually the intention here.

**Comment by Swatinem** [comment]
Oh sorry, I must have mis-read the description.

Considering the following usecase:

Merging tags explicitly set on an event with tags coming from the scope.
As a user, I would expect that the tags I explicitly set on the event to take precedence over the ones in the scope.

Now that I think about this, I notice that the Rust SDK does it the other way around as well:

https://github.com/getsentry/sentry-rust/blob/d35e0e17ce980b84e37287cb935bce8e7bc753e4/sentry-core/src/scope/real.rs#L224-L234

`extend` in Rust is also defined as:

> Inserts all new key-values from the iterator and replaces values with existing keys with new values returned from the iterator.

Considering that this is "broken" in at least one other SDK, I would not really want to bikeshed this for too long and just leave this as is.

**Comment by flub** [comment]
> Merging tags explicitly set on an event with tags coming from the scope. As a user, I would expect that the tags I explicitly set on the event to take precedence over the ones in the scope.

This function can still be used to do it either way round, it is not used yet and there is no restriction on who's src and who's dst here.

File: src/sentry_value.c (hunk 1 above)
Comment (flub) [reply src/sentry_value.c:832]: The intention, and I think also description, is that src values overwrite dst values.  So i don't think there's a need to check `dst_val` for null here?

File: src/sentry_value.h [hunk 2]
```diff
@@ -96,6 +97,18 @@ sentry_value_t sentry__value_clone(sentry_value_t value);
 int sentry__value_append_bounded(
     sentry_value_t value, sentry_value_t v, size_t max);
 
+/**
+ * Deep-merges two objects into each other.
```
Comment (loewenheim) [inline src/sentry_value.h:101]: Nit: This does not actually merge the objects "into each other", it merges `src` into `dst`.

**Review by loewenheim — APPROVED** [review]

**Comment by flub** [comment]
Are there any other places that this should be used?

File: src/sentry_value.c [hunk 3]
```diff
@@ -806,6 +807,38 @@ sentry_value_is_null(sentry_value_t value)
     return value._bits == CONST_NULL;
 }
 
+int
+sentry__value_merge_objects(sentry_value_t dst, sentry_value_t src)
+{
+    if (sentry_value_get_type(dst) != SENTRY_VALUE_TYPE_OBJECT
+        || sentry_value_get_type(src) != SENTRY_VALUE_TYPE_OBJECT
+        || sentry_value_is_frozen(dst)) {
+        return 1;
```
Comment (Swatinem) [inline src/sentry_value.c:816]: oh, I think the problem with the test assertions is that the event (aka `dst`) does not have any tags/extra set at all, so this short circuits.

Comment (flub) [reply src/sentry_value.c:816]: hum, maybe it should be a graceful no-op if the src is null?

Comment (loewenheim) [reply src/sentry_value.c:816]: I think that's justifiable. What should happen if `dst` is null? Replace it wholesale with `src`?

Comment (loewenheim) [reply src/sentry_value.c:816]: You could also keep this function's semantics "pure" and add a wrapper that handles the cases of `src` or `dst` being null.

Comment (flub) [reply src/sentry_value.c:816]: yeah, i think `src` being null is fine in this function.  but for `dst` being null there's not much you can do as the caller likely has other shenanigans to deal with

File: src/sentry_value.c [hunk 4]
```diff
@@ -1,3 +1,4 @@
+#include "sentry.h"
```
Comment (Swatinem) [inline src/sentry_value.c:1]: Remove this include. this is being pulled in via `sentry_boot.h`, and that should also magically fix your compile error.

Don’t ask me to explain *exactly* whats going on. Just that in C include order matters, and defines that you set have an effect on includes, and `sentry_boot.h` sets some defines that are needed so the Windows SDK does not break on microsofts own compiler. Fun times! Rewrite it in Rust!

File: src/sentry_value.h [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #ifndef SENTRY_VALUE_H_INCLUDED
 #define SENTRY_VALUE_H_INCLUDED
 
+#include "sentry.h"
```
Comment (Swatinem) [inline src/sentry_value.h:4]: same here I guess… Not sure why you added this in the first place?

File: src/sentry_value.c (hunk 4 above)
Comment (flub) [reply src/sentry_value.c:1]: oh, thanks!

File: src/sentry_value.h (hunk 5 above)
Comment (flub) [reply src/sentry_value.h:4]: i didn't, clangd did

**Comment by relaxolotl** [comment]
fixed the build, addressed a minor bug, made some minor tweaks to the scope merging by using available macros, and added some additional usages of the merge in places that could use it. the actual merging implementation was not touched.

would appreciate a second look at the changes just to be safe.

File: tests/unit/test_value.c [hunk 6]
```diff
@@ -249,6 +249,7 @@ SENTRY_TEST(value_object_merge)
     TEST_CHECK_INT_EQUAL(sentry_value_as_int32(a), 1);
     TEST_CHECK_INT_EQUAL(sentry_value_as_int32(b), 20);
     TEST_CHECK_INT_EQUAL(sentry_value_as_int32(c), 30);
+
```
Comment (flub) [inline tests/unit/test_value.c:252]: huh, surely this commit can't be doing anything?  I only inserts two blank lines?

**Pull request merged**